internal/clients/init.go

```diff
@@ -155,70 +153,82 @@ func checkServiceAvailableViaRegistry(serviceId string) bool {
-
-	params.Path = clients.ApiAddressableRoute
-	params.Url = metaAddr + params.Path
-	common.AddressableClient = metadata.NewAddressableClient(params, endpoint)
-
-	params.Path = clients.ApiDeviceRoute
-	params.Url = metaAddr + params.Path
-	common.DeviceClient = metadata.NewDeviceClient(params, endpoint)
-
-	params.Path = clients.ApiDeviceServiceRoute
-	params.Url = metaAddr + params.Path
-	common.DeviceServiceClient = metadata.NewDeviceServiceClient(params, endpoint)
-
-	params.Path = clients.ApiDeviceProfileRoute
-	params.Url = metaAddr + params.Path
-	common.DeviceProfileClient = metadata.NewDeviceProfileClient(params, endpoint)
-
-	params.Path = clients.ApiConfigRoute
-	params.Url = metaAddr
-	common.MetadataGeneralClient = general.NewGeneralClient(params, endpoint)
-
-	params.Path = clients.ApiProvisionWatcherRoute
-	params.Url = metaAddr + params.Path
-	common.ProvisionWatcherClient = metadata.NewProvisionWatcherClient(params, endpoint)
+	common.AddressableClient = metadata.NewAddressableClient(
+		urlclient.NewMetadata(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiAddressableRoute,
+		),
+	)
+
+	common.DeviceClient = metadata.NewDeviceClient(
+		urlclient.NewMetadata(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiDeviceRoute,
+		),
+	)
+
+	common.DeviceServiceClient = metadata.NewDeviceServiceClient(
+		urlclient.NewMetadata(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiDeviceServiceRoute,
+		),
+	)
+
+	common.DeviceProfileClient = metadata.NewDeviceProfileClient(
+		urlclient.NewMetadata(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiDeviceProfileRoute,
+		),
+	)
+
+	common.MetadataGeneralClient = general.NewGeneralClient(
+		urlclient.NewMetadata(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiConfigRoute,
+		),
+	)
+
+	common.ProvisionWatcherClient = metadata.NewProvisionWatcherClient(
+		urlclient.NewMetadata(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiProvisionWatcherRoute,
+		),
+	)
 
 	// initialize Core Data clients
-	params.ServiceKey = clients.CoreDataServiceKey
-
-	params.Path = clients.ApiEventRoute
-	params.Url = dataAddr + params.Path
-	common.EventClient = coredata.NewEventClient(params, endpoint)
-
-	params.Path = common.APIValueDescriptorRoute
-	params.Url = dataAddr + params.Path
-	common.ValueDescriptorClient = coredata.NewValueDescriptorClient(params, endpoint)
-
-	if isRegistry {
+	common.EventClient = coredata.NewEventClient(
+		urlclient.NewData(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiEventRoute,
+		),
+	)
+
+	common.ValueDescriptorClient = coredata.NewValueDescriptorClient(
+		urlclient.NewData(
+			ctx,
+			common.RegistryClient,
+			waitGroup,
+			common.APIValueDescriptorRoute,
+		),
+	)
+
+	if common.RegistryClient != nil {
 		// wait for the first endpoint discovery to make sure all clients work
 		waitGroup.Wait()
```

Comment:
Registry no longer has a long running func. That has moved to the config provider so if just using -r, this will wait forever.

Bootstrap's Run() has the needed call to Wait() here:
https://github.com/edgexfoundry/go-mod-bootstrap/blob/master/bootstrap/bootstrap.go#L194

So this call to Wait() needs to be removed.